### DISCUSSION TITLE
How to refresh installed version information of casks with `auto_updates`?

### DISCUSSION THREAD

**Discussion by woj-tek** [Homebrew/discussions#4868] [Everyday usage]
### Output of `brew config`

```shell
$ brew config
HOMEBREW_VERSION: 4.1.17
ORIGIN: https://github.com/Homebrew/brew
HEAD: 35746e0a6ba6c3c5cfe56d99f79d9ec9f52ee15f
Last commit: 6 days ago
Core tap JSON: 28 Oct 09:34 UTC
HOMEBREW_PREFIX: /opt/homebrew
HOMEBREW_CASK_OPTS: []
HOMEBREW_EDITOR: nvim
HOMEBREW_GITHUB_API_TOKEN: set
HOMEBREW_MAKE_JOBS: 10
Homebrew Ruby: 2.6.10 => /System/Library/Frameworks/Ruby.framework/Versions/2.6/usr/bin/ruby
CPU: 10-core 64-bit arm_firestorm_icestorm
Clang: 15.0.0 build 1500
Git: 2.42.0 => /opt/homebrew/bin/git
Curl: 8.1.2 => /usr/bin/curl
macOS: 14.0-arm64
CLT: 15.0.0.0.1.1694021235
Xcode: 15.0.1
Rosetta 2: false
```

### Output of `brew doctor`

```shell
$ brew doctor
Please note that these warnings are just used to help the Homebrew maintainers
with debugging if you file an issue. If everything you use Homebrew for is
working fine: please don't worry or file an issue; just ignore this. Thanks!

Warning: Some installed kegs have no formulae!
This means they were either deleted or installed manually.
You should find replacements for the following formulae:
  docker-machine-completion
  docker-compose-completion

Warning: Some installed formulae are deprecated or disabled.
You should find replacements for the following formulae:
  dotnet
  mysql-client@5.7
  openssl@1.1

Warning: You have the following deprecated, official taps tapped:
  Homebrew/homebrew-cask-drivers
Untap them with `brew untap`.

Warning: Putting non-prefixed coreutils in your path can cause GMP builds to fail.

Warning: You have unlinked kegs in your Cellar.
Leaving kegs unlinked can lead to build-trouble and cause formulae that depend on
those kegs to fail to run properly once built. Run `brew link` on these:
  python-certifi
  python-pytz
  python-packaging
  python-click
  pygments
  docutils
  dart
  dosbox-x
  handbrake
  pyyaml
  python-typing-extensions
  virtualenv
```

### Description of issue

Is there a way to refresh installed version information of installed casks so that when using `brew outdated --cask --greedy` I would actually see relevant upgrades? Currently, even though particular app upgraded itself ("auto_update") brew lists some ancient version that was installed when `brew install --cask <cask>` was executed.

It would help to see what apps actually have upgrades available without updating everything (it turned out today that some apps I don't want to upgrade…)

**Comment by SMillerDev** [comment]
There is not, since there is no specific way to request the current version of whatever the cask installed.

**Reply by woj-tek** [reply]
Hmm... plist?

Excerpt from `brew outdated --cask --greedy`:

```
keka (1.2.50) != 1.3.5
vlc (3.0.16) != 3.0.19
heroic (2.9.1) != 2.9.2
iterm2 (3.4.15) != 3.4.21
thunderbird (91.4.1) != 115.4.1
```

The problematic thing is locating where the app is installed but it seems to be defined with [`app` keyword](https://docs.brew.sh/Cask-Cookbook#stanza-app) (not sure if it's mandatory):

[`app "VLC.app"`](https://github.com/Homebrew/homebrew-cask/blob/70e21e9a47f4e7efe97b3107d4b2aa53865f9936/Casks/v/vlc.rb#L21C3-L21C16)
[`app "Heroic.app"`](https://github.com/Homebrew/homebrew-cask/blob/70e21e9a47f4e7efe97b3107d4b2aa53865f9936/Casks/h/heroic.rb#L15)
[`app "iTerm.app"`](https://github.com/Homebrew/homebrew-cask/blob/70e21e9a47f4e7efe97b3107d4b2aa53865f9936/Casks/i/iterm2.rb#L46C2-L46C2)
[`app "Keka.app"`](https://github.com/Homebrew/homebrew-cask/blob/70e21e9a47f4e7efe97b3107d4b2aa53865f9936/Casks/k/keka.rb#L19)
(those I checked)

and knowing where it's installed it's easy to get the current version from the plist:

```
$ plutil -p /Applications/VLC.app/Contents/Info.plist | grep CFBundleShortVersionString
  "CFBundleShortVersionString" => "3.0.18"
```

```
$ plutil -p /Applications/Heroic.app/Contents/Info.plist | grep CFBundleShortVersionString
  "CFBundleShortVersionString" => "2.9.1"
```

```
$ plutil -p /Applications/Keka.app/Contents/Info.plist | grep CFBundleShortVersionString
  "CFBundleShortVersionString" => "1.3.5"
```

```
$ plutil -p /Applications/Thunderbird.app/Contents/Info.plist | grep CFBundleShortVersionString
  "CFBundleShortVersionString" => "115.4.1"
```

So it seems quite doable (at least for those that include `app` keyword)?

**Reply by SMillerDev** [reply]
That's the problem, that isn't always the version that is represented in the cask or by upstream though. 

Additionally, there are a lot of pkg and even font casks out there too.

In the end, I'm sure people would happily review it if you made a pull request for it, but it's not there currently because it's complicated and not a problem for most people .

**Reply by woj-tek** [reply]
> That's the problem, that isn't always the version that is represented in the cask or by upstream though.

What do you mean?

> Additionally, there are a lot of pkg and even font casks out there too.

I would imagine, though this _could_ be limited to regular apps that would support it.

> In the end, I'm sure people would happily review it if you made a pull request for it, but it's not there currently because it's complicated and not a problem for most people .

Hmm... If I were to take a jab at it, what would be the best point to include it so it would have the highest chance of being merged? I would imagine that `brew outdated --cask --greedy`, for packages that have `app` definition *and* the version can be obtained, would use this mechanism to check the current/installed version -- would that be correct approach?

**Reply by SMillerDev** [reply]
>> That's the problem, that isn't always the version that is represented in the cask or by upstream though.
> What do you mean?

That key is just metadata. Nothing is stopping a vendor from making that key 5.0 and communicating everywhere that the version is 51. That would mean the cask is representing 51 and it would never match.

**Reply by SMillerDev** [reply]
> Hmm... If I were to take a jab at it, what would be the best point to include it so it would have the highest chance of being merged? I would imagine that brew outdated --cask --greedy, for packages that have app definition and the version can be obtained, would use this mechanism to check the current/installed version -- would that be correct approach?

I'd say to include it in `brew info` first so we can validate the version checking independently of other functionality.

**Reply by woj-tek** [reply]
> I'd say to include it in `brew info` first so we can validate the version checking independently of other functionality.

Thanks for the pointer!
